app/controllers/devise_token_auth/concerns/set_user_by_token.rb

```diff
@@ -115,54 +115,61 @@ def update_auth_header
       response.headers.merge!(auth_header)
 
     else
-
-      ensure_pristine_resource do
-        # Lock the user record during any auth_header updates to ensure
-        # we don't have write contention from multiple threads
-        @resource.with_lock do
-          # should not append auth header if @resource related token was
-          # cleared by sign out in the meantime
-          return if @used_auth_by_token && @resource.tokens[@client_id].nil?
-
-          # determine batch request status after request processing, in case
-          # another processes has updated it during that processing
-          @is_batch_request = is_batch_request?(@resource, @client_id)
-
-          auth_header = {}
-
-          # extend expiration of batch buffer to account for the duration of
-          # this request
-          if @is_batch_request
-            auth_header = @resource.extend_batch_buffer(@token, @client_id)
-
-            # Do not return token for batch requests to avoid invalidated
-            # tokens returned to the client in case of race conditions.
-            # Use a blank string for the header to still be present and
-            # being passed in a XHR response in case of
-            # 304 Not Modified responses.
-            auth_header[DeviseTokenAuth.headers_names[:"access-token"]] = ' '
-            auth_header[DeviseTokenAuth.headers_names[:"expiry"]] = ' '
-
-          # update Authorization response header with new token
-          else
-            auth_header = @resource.create_new_auth_token(@client_id)
-          end
-
-          # update the response header
-          response.headers.merge!(auth_header)
-
-        end # end lock
-      end # end ensure_pristine_resource
+      unless @resource.valid?
+        @resource = resource_class.find(@resource.id) # errors are keep after reload
+        # if we left the model in a bad state, something is wrong in our app
+        raise ActiveRecord::RecordInvalid unless @resource.valid?
```

Comment:
`ActiveRecord`'s and `Mongoid`'s `.find` method return exceptions with different names. I think we should avoid such exceptions.

Where will you catch `ActiveRecord::RecordInvalid` error?

Reply:
As the readme states, this helps the developer fix his applicacion, the user was saved invalid somewhere (outside the gem) and now it doesn't let store the tokens because it's invalid. It happened to me sometimes. We can throw a custom exception that says the model is invalid, wdyt?

Reply:
I'm OK with custom exception.

Does front-end client understand that something gone wrong (returned error message)?

Reply:
This should lead to 500, it's a bug in the developer's software, the model is invalid, so it isn't expected and should be fixed